jax/_src/third_party/scipy/special.py

```diff
@@ -0,0 +1,315 @@
+    fresnl_sd = jnp.array([
+      +1.00000000000000000000e0,
+      +2.81376268889994315696e2,
+      +4.55847810806532581675e4,
+      +5.17343888770096400730e6,
+      +4.19320245898111231129e8,
+      +2.24411795645340920940e10,
+      +6.07366389490084639049e11,
+    ])
+
+    fresnl_cn = jnp.array([
+      -4.98843114573573548651e-8,
+      +9.50428062829859605134e-6,
+      -6.45191435683965050962e-4,
+      +1.88843319396703850064e-2,
+      -2.05525900955013891793e-1,
+      +9.99999999999999998822e-1,
+    ])
+
+    fresnl_cd = jnp.array([
+      +3.99982968972495980367e-12,
+      +9.15439215774657478799e-10,
+      +1.25001862479598821474e-7,
+      +1.22262789024179030997e-5,
+      +8.68029542941784300606e-4,
+      +4.12142090722199792936e-2,
+      +1.00000000000000000118e0,
+    ])
+
+    fresnl_fn = jnp.array([
+      +4.21543555043677546506e-1,
+      +1.43407919780758885261e-1,
+      +1.15220955073585758835e-2,
+      +3.45017939782574027900e-4,
+      +4.63613749287867322088e-6,
+      +3.05568983790257605827e-8,
+      +1.02304514164907233465e-10,
+      +1.72010743268161828879e-13,
+      +1.34283276233062758925e-16,
+      +3.76329711269987889006e-20,
+    ])
+
+    fresnl_fd = jnp.array([
+      +1.00000000000000000000e0,
+      +7.51586398353378947175e-1,
+      +1.16888925859191382142e-1,
+      +6.44051526508858611005e-3,
+      +1.55934409164153020873e-4,
+      +1.84627567348930545870e-6,
+      +1.12699224763999035261e-8,
+      +3.60140029589371370404e-11,
+      +5.88754533621578410010e-14,
+      +4.52001434074129701496e-17,
+      +1.25443237090011264384e-20,
+    ])
+
+    fresnl_gn = jnp.array([
+      +5.04442073643383265887e-1,
+      +1.97102833525523411709e-1,
+      +1.87648584092575249293e-2,
+      +6.84079380915393090172e-4,
+      +1.15138826111884280931e-5,
+      +9.82852443688422223854e-8,
+      +4.45344415861750144738e-10,
+      +1.08268041139020870318e-12,
+      +1.37555460633261799868e-15,
+      +8.36354435630677421531e-19,
+      +1.86958710162783235106e-22,
+    ])
+
+    fresnl_gd = jnp.array([
+      +1.00000000000000000000e0,
+      +1.47495759925128324529e0,
+      +3.37748989120019970451e-1,
+      +2.53603741420338795122e-2,
+      +8.14679107184306179049e-4,
+      +1.27545075667729118702e-5,
+      +1.04314589657571990585e-7,
+      +4.60680728146520428211e-10,
+      +1.10273215066240270757e-12,
+      +1.38796531259578871258e-15,
+      +8.39158816283118707363e-19,
+      +1.86958710162783236342e-22,
+    ])
+
+  x = jnp.abs(xxa)
+
+  x2 = x * x
+
+  # Infinite x values
+  s_inf = c_inf = 0.5
+
+  # Small x values
+  t = x2 * x2
+
+  if single_precision:
+    s_small = x * x2 * jnp.polyval(fresnl_sn, t)
+    c_small = x * jnp.polyval(fresnl_cn, t)
+  else:
+    s_small = x * x2 * jnp.polyval(fresnl_sn[:6], t) / jnp.polyval(fresnl_sd[:7], t)
+    c_small = x * jnp.polyval(fresnl_cn[:6], t) / jnp.polyval(fresnl_cd[:7], t)
+
+  # Large x values
+
+  sinpi, cospi = sincospi(x2 / 2)
```

Comment:
```suggestion
  sinpi, cospi = sincospisquaredhalf(x)
```
that ought to resolve https://github.com/google/jax/pull/22843#discussion_r1722119910, wdyt @jakevdp ?

Reply:
Good point, I do not need `sincospi` anymore then ^^'
But the problem persists.

Reply:
Which problem persists?

Reply:
> Which problem persists?

The tests failure on gradients for single precision.

Reply:
What's the issue now? It seems like we solved the initial problem, which was due to `x*x` losing the phase information.

Reply:
I hope you don't mind the screenshot 
![image](https://github.com/user-attachments/assets/24363bce-16f2-4dc4-b6f1-36db0404e8da)

The captured stdout comes from an added `print(f"{x = }")` line at the start of the `fresnel` function body.

Reply:
OK – do you understand why this is going wrong? If we understand it and cannot fix it for fundamental reasons, we should disable the test.

If we don't understand it, we should not disable the test, because it may indicate a bug in the implementation.

Please let me know if you'd like to investigate this, or if you'd like me to fork your branch and take a look.

Reply:
So I make a small experimental setup that tries to emulate (visually) what `check_grads` does, see the code below.

From the below results (first for `float32`, then for `float64`), I think of the following:

- Using FD as "ground truth" makes little sense when x is large, as the rapid oscillations, with very small amplitude, are hardly possible to capture with FD (changing the `EPS` value does impact the result, but not much);
- For such high values, we will eventually have an issue with single-precision's accuracy, and I don't think we can do anything but acknowledge it (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).

I was a bit surprised that my FD estimate would fail on double-precision too, but variations are very small, so this may explain the "issue" of FD being always zero in that range.

Feel free to explore the code and change the `MIN` value to test different ranges.

Let me know what you think of that :-)

![float32](https://github.com/user-attachments/assets/fb1adea6-cbe9-4b28-a99e-57c680fedd08)

![float64](https://github.com/user-attachments/assets/be706d6c-2bbe-4aa6-87b1-80f679b185b9)



```python
import jax
import jax.numpy as jnp
import jax.scipy.special as jsp
import matplotlib.pyplot as plt
import numpy as np
import scipy.special as sp

dtype = np.float64

MIN = 30000

with jax.experimental.enable_x64(dtype == np.float64):
  x = np.linspace(MIN, MIN + 1, 201, dtype=dtype)
  jx = jnp.asarray(x)
  s, c = sp.fresnel(x)
  js, jc = jsp.fresnel(x)
  djs = jax.vmap(jax.grad(lambda x: jsp.fresnel(x)[0]))(jx)
  djc = jax.vmap(jax.grad(lambda x: jsp.fresnel(x)[1]))(jx)

  fig, (ax1, ax2) = plt.subplots(2, 1, figsize=(15, 10))

  color = "tab:red"
  ax1.set_xlabel('x')
  ax1.set_ylabel("S(x)", color=color)
  ax1.plot(x, s, color=color, label="SciPy")
  ax1.plot(x, js, color="k", linestyle="dotted", alpha=0.5, label="JAX")
  ax1.tick_params(axis="y", labelcolor=color)
  ax1.legend()

  ax1b = ax1.twinx()

  colorb = "tab:blue"
  ax1b.set_ylabel("C(x)", color=colorb)
  ax1b.plot(x, c, color=colorb, label="SciPy")
  ax1b.plot(x, jc, color="k", linestyle="dashed", alpha=0.5, label="JAX")
  ax1b.tick_params(axis="y", labelcolor=colorb)
  ax1b.legend()

  EPS = 1e-3
  s_pos, c_pos = jsp.fresnel(jx + EPS)
  s_neg, c_neg = jsp.fresnel(jx - EPS)

  ds = (s_pos - s_neg) * (0.5 / EPS)
  dc = (c_pos - c_neg) * (0.5 / EPS)

  ax2.set_xlabel('x')
  ax2.set_ylabel("dS(x)", color=color)
  ax2.plot(x, ds, color=color, label="FD")
  ax2.plot(x, djs, color="k", linestyle="dotted", alpha=0.5, label="JAX")
  ax2.tick_params(axis="y", labelcolor=color)
  ax2.legend()

  ax2b = ax2.twinx()

  ax2b.set_ylabel("dC(x)", color=colorb)
  ax2b.plot(x, dc, color=colorb, label="FD")
  ax2b.plot(x, djc, color="k", linestyle="dashed", alpha=0.5, label="JAX")
  ax2b.tick_params(axis="y", labelcolor=colorb)
  ax2b.legend()

  fig.tight_layout()
  plt.show()
```

Reply:
Notice that
```python
>>> numpy.nextafter(numpy.float32(30000), numpy.float32(numpy.inf))
30000.002
```
that is, `EPS=1e-3` is too small for single precision large values such as 30000 as `jsp.fresnel(x + EPS)` is very likely equal to `jsp.fresnel(x)` because `x + EPS` is likely equal to `x`. Try `EPS=0.01` but as you said, `jsp.frensel(x)` at the given range is rapidly changing (recall `S(x)->const - cos(x*x)/(2*x) + ...`, `C(x)->const + sin(x*x)/(2*x) + ...` as `x->oo` where `const -> 0.6267..`) so that computing FD may have a little sense indeed. On the other hand, try also
```python
x = np.linspace(MIN, MIN + 0.01, 201, dtype=dtype)
```
to see if you are able to get a smooth graph at all.

Why `dC(MIN)` is equal to 1? Looking at the C(x) graph, I would expect it to be 0.

Also, what is the upper limit for `MIN` to get reasonable results?

Reply:
> Notice that
> 
> ```python
> >>> numpy.nextafter(numpy.float32(30000), numpy.float32(numpy.inf))
> 30000.002
> ```
> 
> that is, `EPS=1e-3` is too small for single precision large values such as 30000 as `jsp.fresnel(x + EPS)` is very likely equal to `jsp.fresnel(x)` because `x + EPS` is likely equal to `x`. Try `EPS=0.01` but as you said, `jsp.frensel(x)` at the given range is rapidly changing (recall `S(x)->const - cos(x*x)/(2*x) + ...`, `C(x)->const + sin(x*x)/(2*x) + ...` as `x->oo` where `const -> 0.6267..`) so that computing FD may have a little sense indeed.

I agree that `1e-3` may not be appropriate for that range of values, but I chose this value to be the same as the one used by `check_grads`. But there is no best `EPS` value, and using FD might just be bad for this function (for large input values).

> On the other hand, try also
> 
> ```python
> x = np.linspace(MIN, MIN + 0.01, 201, dtype=dtype)
> ```
> 
> to see if you are able to get a smooth graph at all.

![float32_1e-1eps](https://github.com/user-attachments/assets/cd88f5be-5c6a-4d65-b0aa-7eff7247d5b9)

> 
> Why `dC(MIN)` is equal to 1? Looking at the C(x) graph, I would expect it to be 0.
> 
> Also, what is the upper limit for `MIN` to get reasonable results?

Humm good remark, I need to double check that :-)

Reply:
> * (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).

In that case, we should note this in the function's docstring and ensure that our tests cover the supported range.

Reply:
> Thanks for digging in! This all makes sense.
> 
> > * (the accuracy of the Fresnel function is usually guaranteed for small values, e.g., `|x| < 5...10`, not for large ones, see the Cephes documentation).
> 
> In that case, we should note this in the function's docstring and ensure that our tests cover the supported range.

Would reducing the scale factor of the random normal variable used for generating test samples be sufficient for you?